---
title: "Differential Gene Expression Analysis"
author: "Expression_Quant"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: yes
---



```{r setup, include=FALSE}
# The following setting is important, do not omit.
knitr::opts_chunk$set(echo = TRUE,
                      fig.height = 8,
                      fig.width = 12,
                      dev = c("png", "pdf"),
                      dpi = 1000)
options(scipen = 999) # Prevent scientific notation
rm(list = ls()) # Clear environment
```

*Note: Open scripts.Rproj first, then script. To easily use relative paths, click the down button next to knit and then click "Knit Directory --> Project Directory". This should make loading and saving files much easier.*



# Differential Gene Expression Analysis

Perform Differential Gene Expression (DGE) analysis using a read count table (usually from Salmon) and a sample metadata sheet (user made).
<span style="color:red">Red text</span> indicates regions that require the user to modify.
Regardless, the user should check over all code blocks to ensure that everything is running correctly.



## 1. Load packages

Load packages
```{r load_packages, message=FALSE, warning=FALSE}
library(edgeR, quietly = TRUE) #edgeR-v3.30.3
library(vegan, quietly = TRUE)
library(Dune, quietly = TRUE)
library(ggplot2, quietly = TRUE) #ggplot2-v3.3.5
library(tidyverse, quietly = TRUE) #tidyverse-v1.3.1
library(ComplexHeatmap, quietly = TRUE)
library(DESeq2, quietly = TRUE)
library(genefilter, quietly = TRUE)
```



## 2. Set variables

<span style="color:red">**Change file names and conditions where appropriate.**</span>
```{r variables}
# Input unfiltered data
treatmentinfo.file <- "samples.tsv"
gcount.file <- "salmon.numreads.tsv"
# Output DEG results
DGE_results.file <- "salmon.numreads.DGE_results.tsv"
DGE_samples.file <- "salmon.numreads.DGE_samples.tsv"

# Treatment conditions and variables to consider
condition <- "treatment"
treatments.to.compare <- c("HTAC", "ATAC")
treatment.colors <- c("ATAC" = "#1b9e77", "HTAC" = "#d95f02")

# pOverA gene filtering thresholds (only process genes with over these cutoffs)
pOverA.cutoff.P <- 0.8 # >80% of samples
pOverA.cutoff.A <- 10  # >10 read counts

# DESeq2 results filtering thresholds (only report results below these cutoffs)
lfcThreshold.cutoff <- log2(1.5) # Log threshold to use for DESeq2 results filtering
padj.cutoff <- 0.05 # Adjusted p-value threshold to use for DESeq2 results filtering
```



## 3. Load, clean, and pre-processing datasets

Load the input file containing the treatment information.
```{r load_treatmentinfo}
treatmentinfo <- read.csv(treatmentinfo.file, header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM") # Read in file
head(treatmentinfo)

# Check we have the right column names
headers <- c("sample_id", condition)
if( all(headers %in% colnames(treatmentinfo)) ){
  cat(paste(treatmentinfo.file, "has the required columns:", paste(headers, collapse=', ')))
} else {
  stop(paste(treatmentinfo.file, "is missing required columns:", paste(headers, collapse=', ')))
}
```


Filter `treatmentinfo` to just the samples that we want to analyze.
<span style="color:red">**Adjust filtering as required.**</span>
```{r filter_treatmentinfo}
treatmentinfo <- treatmentinfo %>%
  filter(!!rlang::sym(condition) %in% treatments.to.compare & timepoint=="8") %>%
  mutate({{condition}} := factor(!!rlang::sym(condition), levels = treatments.to.compare))
head(treatmentinfo)
```


Load the input file containing the gene count matrix.
```{r load_gcount}
gcount <- as.data.frame(read.csv(gcount.file, header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM")) # Read in file

# Check we have the right column names
headers <- c("Name", treatmentinfo$sample_id)
if( all(headers %in% colnames(gcount)) ){
  cat(paste(gcount.file, "has the required columns:", paste(headers, collapse=', ')))
} else {
  stop(paste(gcount.file, "is missing required columns:", paste(headers, collapse=', ')))
}

# Cleanup gcount data
gcount <- gcount %>% 
  column_to_rownames("Name") %>% # Makes "Name" column rownames
  round() %>% # Round
  select(treatmentinfo$sample_id) # Select just columns in filtered treatmentinfo file

# View dataset attributes
d <- dim(gcount); print(paste("rows:",d[[1]]," columns:",d[[2]], sep=''))
head(gcount)[,1:3]
```


Filter `gcount` to just genes with enough data across columns (exclude genes which are mostly zeros)
```{r filter_gcount}
# Create filter for the counts data
filt <- filterfun(pOverA(pOverA.cutoff.P, pOverA.cutoff.A))
gfilt <- genefilter(gcount, filt)
# Identify genes to keep by count filter
keep <- gcount[gfilt,]
# Identify gene lists
keep <- rownames(keep)
# Gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% keep),])
d <- dim(gcount);      print(paste("gcount      - rows:",d[[1]]," columns:",d[[2]], sep='')) # Before filtering
d <- dim(gcount_filt); print(paste("gcount_filt - rows:",d[[1]]," columns:",d[[2]], sep='')) # After filtering
```


Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at the column listed in the `condition` variable.
```{r create_DESeq2_dataset}
gdds <- DESeqDataSetFromMatrix(countData = as.data.frame(gcount_filt),
                               colData = treatmentinfo,
                               design = as.formula(paste("~",condition, sep=''))
                               )
```



## 4. Differential Gene Expression Analysis

### Run DE analysis

Run differential expression test using a Wald model. 
```{r run_DESeq2}
DEG <- DESeq(gdds)
```


Explore significant *p*-values for condition pairs
```{r extract_results}
# Extract DESeq2 results
DEG.results <- results(DEG, contrast= c(condition, treatments.to.compare), lfcThreshold=lfcThreshold.cutoff)
head(DEG.results)

# Filter results
DEGs <- as.data.frame(subset(DEG.results, padj<padj.cutoff))
nrow(DEGs)

# Order p-values by smallest value first
results.ordered <- order(DEGs$padj)

# Make row names a column before writing to file.
DEGs$gene_id  <- rownames(DEGs)
rownames(DEGs) <- NULL
DEGs <- relocate(DEGs, "gene_id", .before="baseMean")

# Write filtered DGE results and samples
write.table(DEGs, DGE_results.file, sep='\t', row.names=F, quote=F)
write.table(treatmentinfo, DGE_samples.file, sep='\t', row.names=F, quote=F)
```


We will now transform them with cpm for plotting, we will also subset the gene count matrix by the list of DEGs.
```{r transform_to_cpm}
DEGlist <- gdds[DEGs$gene_id, ]
```



Apply a variance stabilizing transformation to minimize effects of small counts and normalize by library size
```{r compute_VST}
# Get size factor for each sample
sf <- estimateSizeFactors(gdds)$sizeFactor
print(paste("Max size factor: ",max(sf), sep=''))
print(paste("Min size factor: ",min(sf), sep=''))

# If above max(size factors) is less than 4, we can use VST.
Gvst <- vst(as.matrix(gcount_filt), blind=TRUE, fitType = "local")
DEGvst <- Gvst[DEGs$gene_id,]
head(DEGvst) # View transformed gene count data
dim(DEGvst)
```


Make a matrix for computing similarity
```{r make_similarity_matrix}
mat <- DEGvst # Make an expression object
mat <- mat - rowMeans(mat) # Difference in expression compared to average across all samples
```



## 5. Plot results

### DEG heatmap

Make a heatmap 
```{r plot_heatmap}
hmTreatment <- subset(treatmentinfo, select=c(condition))
col <- list(); col[[condition]] = treatment.colors # Setup list for heatmap colors
hm_ann_col <- HeatmapAnnotation(df=hmTreatment, col = col) # Make dataframe for column naming
dend = cluster_within_group(mat, hmTreatment[[condition]])

DEGheatmap <-  Heatmap(mat,
          cluster_columns = dend,
          column_split = 2,
          name = "Gene expression (vst)",
          show_row_names = F,
          top_annotation = hm_ann_col,
          show_column_names = F,
          row_dend_side = "left",
          column_dend_height = unit(0.5, "in"),
          row_title_side = "right",
          row_title_rot = 0,
          row_dend_reorder = TRUE,
          row_gap = unit(2.5, "mm"),
          border = TRUE,
          column_names_gp =  gpar(fontsize = 10)
        )

draw(DEGheatmap)
```



### Principle components plot of DEGs

```{r plot_PCA}
pca <- prcomp(t(mat)) # Calculate eigengenes
percentVar <- pca$sdev^2/sum(pca$sdev^2) # Save % variation by PC1 and PC2
d <- data.frame(treatmentinfo, PC1 = pca$x[, 1], PC2 = pca$x[, 2])
DEG_PCA <- ggplot(data = d, aes_string(x = "PC1", y = "PC2")) +
  geom_point(size = 4, aes(colour=!!rlang::sym(condition))) +
  xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) +
  ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) +
  coord_fixed() +
  scale_color_manual(values = treatment.colors) +
  theme_bw() + # Set background color
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), # Set major gridlines
        panel.grid.minor = element_blank(), # Set minor gridlines
        axis.line = element_line(colour = "black", size = 0.6), # Set axes color
        plot.background=element_blank(), # Set the plot background
        axis.title = element_text(size = 14), # Axis title size
        axis.text = element_blank()) # Axis text size and view plot
DEG_PCA
```



## 6. Session Info

```{r ressionInfo}
sessionInfo()
```


