---
title: "WGCNA GO Term and KEGG Enrichment Analysis"
author: "Expression_Quant"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: yes
---



```{r setup, include=FALSE}
# The following setting is important, do not omit.
knitr::opts_chunk$set(echo = TRUE,
                      fig.height = 8,
                      fig.width = 12,
                      dev = c("png", "pdf"),
                      dpi = 1000)
options(scipen = 999) # Prevent scientific notation
rm(list = ls()) # Clear environment
```

*Note: Open scripts.Rproj first, then script. To easily use relative paths, click the down button next to knit and then click "Knit Directory --> Project Directory". This should make loading and saving files much easier.*



# GO Term and KEGG Enrichment Analysis

Run GO Term enrichment analysis using results from WGCNA analysis (i.e., after running the `2a-WGCNA.Rmd` script).
<span style="color:red">Red text</span> indicates regions that require the user to modify.
Regardless, the user should check over all code blocks to ensure that everything is running correctly.



## 1. Load packages

Load packages
```{r load_packages, message=FALSE, warning=FALSE}
library(goseq)
library(tidyverse)
library(GSEABase)
library(data.table)
library(ggplot2)
library(cowplot)
library(patchwork)
```



## 2. Set variables

<span style="color:red">**Change file names and conditions where appropriate.**</span>
```{r variables}
# Input unfiltered data
treatmentinfo.file <- "salmon.numreads.WGCNA_samples.tsv"
wgcna_results.file <- "salmon.numreads.WGCNA_results.tsv"
annotation.file <- "Pocillopora_acuta_KBHIv2.pep.GOs.tsv"

# Output DEG results
GO_term_sig.file <- "salmon.numreads.WGCNA_results.GOsig.tsv"
KEGG_sig.file <- "salmon.numreads.WGCNA_results.KEGGsig.tsv"

# Cutoff for significant GO term p-values
GO_enrich_pvalue.cutoff <- 0.05
KEGG_enrich_pvalue.cutoff <- 0.05

# KEGG ID description file - provided with script
KEGG_IDs_Descriptions.file <- "2b-KEGG_IDs_Descriptions.tsv.gz"
```



## 3. Load, clean, and pre-processing datasets

Load the input file containing the treatment information.
```{r load_treatmentinfo}
treatmentinfo <- read.csv(treatmentinfo.file, header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM") # Read in file
head(treatmentinfo)

# Check we have the right column names
headers <- c("sample_id")
if( all(headers %in% colnames(treatmentinfo)) ){
  print(paste(treatmentinfo.file, "has the required columns:", paste(headers, collapse=', ')))
} else {
  stop(paste(treatmentinfo.file, "is missing required columns:", paste(headers, collapse=', ')))
}
```


Load the input file containing WGCNA results.
```{r load_wgcna_results}
wgcna_results <- as.data.frame(read.csv(wgcna_results.file, header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM")) # Read in file

# Check we have the right column names
headers <- c("gene_id")
if( all(headers %in% colnames(wgcna_results)) ){
  print(paste(wgcna_results.file, "has the required columns:", paste(headers, collapse=', ')))
} else {
  stop(paste(wgcna_results.file, "is missing required columns:", paste(headers, collapse=', ')))
}
```


Filter `wgcna_results` to just the module that we want to analyze. 
<span style="color:red">**Adjust filtering as required.**</span>
```{r filter_treatmentinfo}
wgcna_results <- wgcna_results %>%
  filter(moduleColor %in% c("darkred"))
head(wgcna_results)
```


Load the input file containing gene annotations (GO terms, KEGG IDs, and gene/protein lengths).
```{r load_annotations}
# Annotations
annot <- read.csv(annotation.file, header=TRUE, sep="\t", fileEncoding="UTF-8-BOM")

# Check we have the right column names
headers <- c("gene_id", "GO_IDs", "KEGG_IDs", "length")
if( all(headers %in% colnames(annot)) ){
  paste(paste(annotation.file, "has the required columns:", paste(headers, collapse=', ')))
} else {
  stop(paste(annotation.file, "is missing required columns:", paste(headers, collapse=', ')))
}

# Select only relevant information
Go.ref <- subset(annot, select= c(gene_id, length))

# Merge wgcna_results by available annotations
Go.ref <- merge(wgcna_results, Go.ref, by = "gene_id")
Go.ref <- unique(Go.ref)
dim(Go.ref)
```


Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for DEGs.
```{r probability_weighting_function}
# Make ID and length vectors
IDvector <- annot$gene_id
lengthVector <- annot$length

# Get binary list indicating which genes are in WGCNA results and which are not out of all genes
wgcna.genes <- as.integer(annot$gene_id %in% Go.ref$gene_id)
names(wgcna.genes) <- annot$gene_id
print(paste("Number of WGCNA genes:     ", length(wgcna.genes[wgcna.genes == 1]), sep=''))
print(paste("Number of NON WGCNA genes: ", length(wgcna.genes[wgcna.genes == 0]), sep=''))

# Weight vector by length of gene
pwf <- nullp(DEgenes=wgcna.genes, id=IDvector, bias.data=lengthVector)
```


Prepare GO term dataframe.
```{r prepare_GO_term_dataframe}
# Cleanup GO terms and split multiple GO terms per line, into one GO term per line
GO.annot <- annot %>%
  subset(select=c(gene_id, GO_IDs)) %>%
  drop_na() %>% # Remove rows with missing GO terms or gene IDs
  filter(GO_IDs!="-") # Remove rows with missing values indicated by '-'
splitted <- strsplit(as.character(GO.annot$GO_IDs), "; ") # Split into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GO.annot$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) # List all genes with each of their GO terms in a single row
colnames(GO.terms) <- c("gene_id", "GO.ID")

# Cleanup single-line GO term dataframe
GO.terms <- GO.terms %>%
  mutate(GO.ID = gsub(" ", "", GO.ID)) %>% # Remove spaces from GO terms - in case any exist by mistake
  mutate(GO.ID = as.character(GO.ID)) %>%
  mutate(GO.ID = factor(GO.ID, levels=unique(GO.ID))) %>%
  mutate(gene_id = factor(gene_id, levels=unique(gene_id))) %>%
  unique()

# Print stats
print(paste("No rows in 'GO.terms': ", dim(GO.terms)[1], sep=''))
print(paste("Avg GO IDs per gene: ", nrow(GO.terms) / length(unique(GO.terms$gene_id)), sep=''))
head(GO.terms)
```


Find enriched GO terms, "selection-unbiased testing for category enrichment amongst significantly expressed genes for RNA-seq data".
```{r goseq}
GOall <- goseq(pwf, GOref$gene_id, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
```


Find only enriched GO terms that are statistically significant at cutoff.
```{r selected_significant_GOs}
GOall.filtered <- GOall %>% 
  filter(over_represented_pvalue < GO_enrich_pvalue.cutoff | under_represented_pvalue < GO_enrich_pvalue.cutoff) %>% # Filter GO enrichment results by p-value
  arrange(ontology, over_represented_pvalue, -numDEInCat) %>% # Reorder
  mutate(term = factor(term, levels = unique(term))) %>% # Make 'terms' column a factor
  mutate(dir = if_else(over_represented_pvalue < GO_enrich_pvalue.cutoff, 
                       "Over", 
                       if_else(under_represented_pvalue < GO_enrich_pvalue.cutoff, 
                               "Under", 
                               "NULL"
                              )
                      )
        )

# Print stats
print(paste("Number GOs BEFORE sig filtering: ", nrow(GOall), sep=''))
print(paste("Number GOs AFTER sig filtering:  ", nrow(GOall.filtered), sep=''))
print(paste("Number GOs AFTER sig filtering OVER:  ", nrow(GOall.filtered %>% filter(over_represented_pvalue  < GO_enrich_pvalue.cutoff)), sep=''))
print(paste("Number GOs AFTER sig filtering UNDER: ", nrow(GOall.filtered %>% filter(under_represented_pvalue < GO_enrich_pvalue.cutoff)), sep=''))
print(paste("Number sig BP terms: ", nrow(filter(GOall.filtered, ontology=="BP")), sep=''))
print(paste("Number sig MF terms: ", nrow(filter(GOall.filtered, ontology=="MF")), sep=''))
print(paste("Number sig CC terms: ", nrow(filter(GOall.filtered, ontology=="CC")), sep=''))
```


Correct any un-annotated terms/ontologies.
```{r list_unannotated}
NAs.ontology <- GOall.filtered %>% subset(is.na(term))
NAs.ontology
```


Save significant terms.
```{r write_GOall_filtered}
write.table(GOall.filtered, file = GO_term_sig.file, row.names=F, quote=F, sep='\t')
```



## 4. Find GOslim terms

Run GOslim to get broader categories.
```{r run_GOslim}
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") # Get GO database

## BP
BP_GO <- GOall.filtered %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) # Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) # Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) # Save rownames as category

## MF
MF_GO <- GOall.filtered %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) # Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) # Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) # Save rownames as category
```


Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r map_terms}
# Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-function(df, collection, OFFSPRING) { # The command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
  mapped <- lapply(map, intersect, ids(collection)) # Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
  df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) # Add column "go_terms" with matching terms 
  df # Show resulting dataframe
}

# Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
```


Remove duplicate matches, keeping the broader umbrella term
```{r clean_mapped_terms}
# BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") # Filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") # Split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) # List all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") # Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") # Remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) # Rename columns
head(BPslim)

# MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") # Filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") # Split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) # List all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  # Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  # Remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) # Rename columns
head(MFslim)
```


Add back GO enrichment info for each offspring term.
```{r join_GOslim_and_enrichment_results}
GO.BP <- right_join(BPslim, filter(GOall.filtered, ontology=="BP"), by="category")
GO.MF <- right_join(MFslim, filter(GOall.filtered, ontology=="MF"), by="category")
```



## 5. Make heatmap

Plot heatmap of BP and MF GO slim terms.
```{r plot_BP_heatmap}
term_label_text_size <- 6
slim_label_text_size <- 6

BPplot <- GO.BP %>%
  filter(numInCat>5) %>%
  mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
  ggplot(aes(x = dir, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
    facet_grid(slim_term ~ ontology, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
    theme_bw() +
    theme(panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black"),
          strip.text.y = element_text(angle=0, size = slim_label_text_size, face = "bold"),
          strip.text.x = element_text(size = 12, face = "bold"),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size=15),
          axis.text = element_text(size = term_label_text_size), legend.position = "None",
          plot.margin = unit(c(0,1,0,0.25), "cm")
    ); BPplot
```


```{r plot_MF_heatmap}
MFplot <- GO.MF %>%
  filter(numInCat>5) %>%
  mutate(term = fct_reorder(term, -over_represented_pvalue)) %>%
  ggplot(aes(x = dir, y = term)) + 
    geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
    scale_y_discrete(position = "right") +
    facet_grid(slim_term ~ ontology,
               scales = "free_y",
               labeller = label_wrap_gen(width = 10, multi_line = TRUE), 
               switch="y" # Put the y facet strips on the left
              ) +
    theme_bw() +
    theme(panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black"),
          strip.text.y.left = element_text(angle=0, size = slim_label_text_size, face = "bold"),
          strip.text.x = element_text(size = 12, face = "bold"),
          axis.title = element_blank(),
          axis.text = element_text(size = term_label_text_size),
          legend.title = element_text(size = 12),
          legend.text = element_text(size = 11)
    ); MFplot
```


Combined BP and MF plots.
```{r plot_combined_BP_MF_heatmap}
BPplot + MFplot + plot_annotation(tag_levels = "A", tag_suffix = ")") & theme(plot.tag = element_text(size=15, face="bold"))
```



## 6. Kegg enrichment analysis

Select KEGG Orthogroup IDs (KOs) from annotation dataframe.
```{r extract_KO_annots}
# Extract gene_ids and KO from annotation file
KO.terms <- annot %>%
  subset(select=c(gene_id, KEGG_IDs)) %>% # Select columns
  drop_na() %>% # Remove rows with missing GO terms or gene IDs
  filter(KEGG_IDs!="-") # Remove rows with missing values indicated by '-'

# Split multiple KEGG IDs per line into multiple lines
splitted <- strsplit(as.character(KO.terms$KEGG_IDs), "; ") # Split into multiple KEGG IDs
KO.terms <- data.frame(v1 = rep.int(KO.terms$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) # List all genes with each of their KEGG IDs in a single row
colnames(KO.terms) <- c("gene_id", "KEGG_IDs")

KO.terms <- unique(KO.terms)
colnames(KO.terms) <- c("gene_id", "GO.ID")
head(KO.terms)

# Bind KO and GO references
GOKO.terms <- bind_rows(GO.terms, KO.terms)
```


Perform Kegg enrichment with goseq package.
```{r goseq_on_KOs}
KOall <- goseq(pwf, GOref$gene_id, gene2cat=GOKO.terms, test.cats=c("KEGG"), method="Wallenius", use_genes_without_cat=TRUE)
```


Extract significantly enriched KEGG terms.
```{r selected_significant_KOs}
KOall.filtered <- KOall %>% 
  filter(over_represented_pvalue < KEGG_enrich_pvalue.cutoff | under_represented_pvalue < KEGG_enrich_pvalue.cutoff) %>% # Filter GO enrichment results by p-value
  arrange(ontology, over_represented_pvalue, -numDEInCat) %>% # Reorder
  mutate(term = factor(term, levels = unique(term))) %>% # Make 'terms' column a factor
  mutate(dir = if_else(over_represented_pvalue < GO_enrich_pvalue.cutoff, 
                       "Over", 
                       if_else(under_represented_pvalue < GO_enrich_pvalue.cutoff, 
                               "Under", 
                               "NULL"
                              )
                      )
        ) %>%
  mutate(ontology = replace_na(ontology, "KEGG")) %>%
  filter(ontology=="KEGG")

# Print stats
print(paste("Number KEGG IDs BEFORE sig filtering: ", nrow(KOall), sep=''))
print(paste("Number KEGG IDs AFTER sig filtering:  ", nrow(KOall.filtered), sep=''))
print(paste("Number KEGG IDs AFTER sig filtering OVER:  ", nrow(KOall.filtered %>% filter(over_represented_pvalue  < GO_enrich_pvalue.cutoff)), sep=''))
print(paste("Number KEGG IDs AFTER sig filtering UNDER: ", nrow(KOall.filtered %>% filter(under_represented_pvalue < GO_enrich_pvalue.cutoff)), sep=''))
```


Add KO definitions.
```{r add_KEGG_ID_definitions}
# Load definition file
KEGG_IDs_Descriptions <- read.table(KEGG_IDs_Descriptions.file, header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM", quote='') # Read in file

# Prep definition data
KEGG_IDs_Descriptions <- KEGG_IDs_Descriptions %>%
  subset(select=c("D.ID", "D.Description")) %>%
  unique()
colnames(KEGG_IDs_Descriptions) <- c("category", "description")

# Merge with KEGG output
KOall.filtered.descriptions <- unique(left_join(KOall.filtered, KEGG_IDs_Descriptions, by=c("category")))
```


Write output KEGG enrichment files.
```{r write_KEGG_filtered_descriptions}
write.table(KOall.filtered.descriptions, file = KEGG_sig.file, row.names=F, quote=F, sep='\t')
```



## 7. Session Info

```{r ressionInfo}
sessionInfo()
```


